python_modules/dagster/dagster/_core/definitions/time_window_partitions.py

```diff
@@ -54,6 +56,84 @@
 from .partition_key_range import PartitionKeyRange
 
 
+def is_ambiguous_time(dt: datetime):
+    """Returns if a datetime is ambiguous due to DST transitions."""
+    tz = check.not_none(dt.tzinfo)
+    # UTC is never ambiguous
+    if tz.tzname(None) == "UTC":
+        return False
+    offset_before = cast(
+        timedelta,
+        (tz.utcoffset(dt.replace(fold=0)) if dt.fold else tz.utcoffset(dt)),
+    )
+    offset_after = cast(
+        timedelta,
+        (tz.utcoffset(dt) if dt.fold else tz.utcoffset(dt.replace(fold=1))),
+    )
+    return offset_before > offset_after
+
+
+def dst_safe_fmt(fmt: str) -> str:
+    """Adds UTC offset information to a datetime format string to disambiguate timestamps around DST
+    transitions.
+    """
+    if "%z" in fmt:
+        return fmt
+    return fmt + "%z"
+
+
+def dst_safe_strftime(dt: datetime, fmt: str, cron_schedule: str) -> str:
+    """A method for converting a datetime to a string which will append a suffix in cases where
+    the resulting timestamp would be ambiguous due to DST transitions.
+    """
+    time_str = dt.strftime(fmt)
+
+    # if the format already includes a UTC offset, then we don't need to do anything
+    if fmt == dst_safe_fmt(fmt):
+        return time_str
+
+    # only need to handle ambiguous times for cron schedules which repeat every hour
+    if not cron_string_repeats_every_hour(cron_schedule):
+        return time_str
+
+    if is_ambiguous_time(dt) and dt.fold > 0:
+        return dt.strftime(dst_safe_fmt(fmt))
+    return time_str
+
+
+def dst_safe_strptime(date_string: str, tz: str, fmt: str) -> PendulumDateTime:
+    """A method for parsing a datetime created with the dst_safe_strftime() method."""
+    try:
+        # first, try to parse the datetime in the normal format
+        dt = datetime.strptime(date_string, fmt)
+        dst_rule = pendulum.PRE_TRANSITION
+    except ValueError:
+        # if this fails, try to parse the datetime with a UTC offset added
+        dt = datetime.strptime(date_string, dst_safe_fmt(fmt))
+        dst_rule = pendulum.POST_TRANSITION
```

Comment:
i wonder if this is strictly needed now that you have a UTC offset

Reply:
you're right, it wasn't